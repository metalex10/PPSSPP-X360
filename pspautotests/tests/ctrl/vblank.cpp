#include <common.h>
#include <pspctrl.h>
#include <pspdisplay.h>
#include <pspintrman.h>
#include <pspthreadman.h>

extern "C" {
	int sceDisplayIsVsync();
}

static const bool LOG_ACCURATE_TIMING = false;
static const u32 TIME_ROUND_INTERVAL = (1000000 / 60) / 8;
static const u32 HCOUNT_ROUND_INTERVAL = 286 / 2;
static const u32 FRAME_HCOUNT_ROUND_INTERVAL = 5;

static SceCtrlData pad_data[64];
static u32 firstSample = 0;
static int vcountBase;
static int hcountBase;

u32 getLatestSample() {
	u32 latestSample = 0;
	for (int i = 0; i < ARRAY_SIZE(pad_data); ++i) {
		if (latestSample < pad_data[i].TimeStamp) {
			latestSample = pad_data[i].TimeStamp;
		}
	}

	return latestSample;
}

u32 roundUp(u32 v, u32 interval) {
	u32 intervals = (v + interval - 1) / interval;
	return intervals * interval;
}

void logSampleInfo() {
	u32 sample = getLatestSample() - firstSample;
	u32 timeDelta = sceKernelGetSystemTimeLow() - getLatestSample();
	int hcount = sceDisplayGetAccumulatedHcount() - hcountBase;
	int frameHcount = sceDisplayGetCurrentHcount();

	if (!LOG_ACCURATE_TIMING) {
		sample = roundUp(sample, TIME_ROUND_INTERVAL);
		timeDelta = roundUp(timeDelta, TIME_ROUND_INTERVAL);
		hcount = roundUp(hcount, HCOUNT_ROUND_INTERVAL);
		frameHcount = roundUp(frameHcount, FRAME_HCOUNT_ROUND_INTERVAL);
	}

	schedf("[-]   Latest sample at: +%ums, behind: %ums\n", sample, timeDelta);
	schedf("[-]   Vblank: %d (inside: %d), inside vsync: %d\n", sceDisplayGetVcount() - vcountBase, sceDisplayIsVblank(), sceDisplayIsVsync());
	schedf("[-]   Hcount: %d (+%d)\n", hcount, frameHcount);
}

extern "C" void vblankCallback(int no, void *value) {
	schedf("[-] Peek inside interrupt: %08x\n", sceCtrlPeekBufferPositive(pad_data, 64));
	logSampleInfo();
}

extern "C" int main(int argc, char *argv[]) {
	checkpointNext("Init:");
	checkpoint("sceCtrlSetSamplingCycle: %08x", sceCtrlSetSamplingCycle(0));

	sceKernelRegisterSubIntrHandler(PSP_DISPLAY_SUBINT, 1, (void *)vblankCallback, NULL);

	// Clear it out in case we've been running a while.
	sceCtrlReadBufferPositive(pad_data, 64);

	checkpointNext("Vblanks:");

	sceCtrlPeekBufferPositive(pad_data, 64);
	for (int i = 0; i < 64; i++) {
		if (firstSample < pad_data[i].TimeStamp) {
			firstSample = pad_data[i].TimeStamp;
		}
	}

	sceDisplayWaitVblankStart();
	vcountBase = sceDisplayGetVcount();
	hcountBase = sceDisplayGetAccumulatedHcount();

	sceKernelEnableSubIntr(PSP_DISPLAY_SUBINT, 1);
	for (int i = 0; i < 20; ++i) {
		checkpoint("Read outside interrupt: %08x", sceCtrlReadBufferPositive(pad_data, 64));
		logSampleInfo();
	}
	sceKernelDisableSubIntr(PSP_VBLANK_INT, 1);

	sceDisplayWaitVblankStart();
	int avg = (sceDisplayGetAccumulatedHcount() - hcountBase) / 201;
	checkpoint("Possible: %d", avg);

	sceKernelReleaseSubIntrHandler(PSP_DISPLAY_SUBINT, 1);

	return 0;
}
