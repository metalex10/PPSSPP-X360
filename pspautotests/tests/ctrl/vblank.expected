[x] Init:
[x] sceCtrlSetSamplingCycle: 00000000

[x] Vblanks:
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +18747ms, behind: 2083ms
[-]   Vblank: 0 (inside: 0), inside vsync: 0
[-]   Hcount: 143 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +18747ms, behind: 16664ms
[-]   Vblank: 1 (inside: 1), inside vsync: 0
[-]   Hcount: 286 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +35411ms, behind: 2083ms
[-]   Vblank: 1 (inside: 0), inside vsync: 0
[-]   Hcount: 429 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +35411ms, behind: 16664ms
[-]   Vblank: 2 (inside: 1), inside vsync: 0
[-]   Hcount: 572 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +52075ms, behind: 2083ms
[-]   Vblank: 2 (inside: 0), inside vsync: 0
[-]   Hcount: 715 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +52075ms, behind: 16664ms
[-]   Vblank: 3 (inside: 1), inside vsync: 0
[-]   Hcount: 858 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +68739ms, behind: 2083ms
[-]   Vblank: 3 (inside: 0), inside vsync: 0
[-]   Hcount: 1001 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +68739ms, behind: 16664ms
[-]   Vblank: 4 (inside: 1), inside vsync: 0
[-]   Hcount: 1144 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +85403ms, behind: 2083ms
[-]   Vblank: 4 (inside: 0), inside vsync: 0
[-]   Hcount: 1287 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +85403ms, behind: 16664ms
[-]   Vblank: 5 (inside: 1), inside vsync: 0
[-]   Hcount: 1430 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +102067ms, behind: 2083ms
[-]   Vblank: 5 (inside: 0), inside vsync: 0
[-]   Hcount: 1573 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +102067ms, behind: 16664ms
[-]   Vblank: 6 (inside: 1), inside vsync: 0
[-]   Hcount: 1716 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +118731ms, behind: 2083ms
[-]   Vblank: 6 (inside: 0), inside vsync: 0
[-]   Hcount: 1859 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +118731ms, behind: 16664ms
[-]   Vblank: 7 (inside: 1), inside vsync: 0
[-]   Hcount: 2002 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +135395ms, behind: 2083ms
[-]   Vblank: 7 (inside: 0), inside vsync: 0
[-]   Hcount: 2145 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +135395ms, behind: 16664ms
[-]   Vblank: 8 (inside: 1), inside vsync: 0
[-]   Hcount: 2288 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +152059ms, behind: 2083ms
[-]   Vblank: 8 (inside: 0), inside vsync: 0
[-]   Hcount: 2431 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +152059ms, behind: 16664ms
[-]   Vblank: 9 (inside: 1), inside vsync: 0
[-]   Hcount: 2574 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +168723ms, behind: 2083ms
[-]   Vblank: 9 (inside: 0), inside vsync: 0
[-]   Hcount: 2717 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +168723ms, behind: 16664ms
[-]   Vblank: 10 (inside: 1), inside vsync: 0
[-]   Hcount: 2860 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +185387ms, behind: 2083ms
[-]   Vblank: 10 (inside: 0), inside vsync: 0
[-]   Hcount: 3003 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +185387ms, behind: 16664ms
[-]   Vblank: 11 (inside: 1), inside vsync: 0
[-]   Hcount: 3146 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +202051ms, behind: 2083ms
[-]   Vblank: 11 (inside: 0), inside vsync: 0
[-]   Hcount: 3289 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +202051ms, behind: 16664ms
[-]   Vblank: 12 (inside: 1), inside vsync: 0
[-]   Hcount: 3432 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +218715ms, behind: 2083ms
[-]   Vblank: 12 (inside: 0), inside vsync: 0
[-]   Hcount: 3575 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +218715ms, behind: 16664ms
[-]   Vblank: 13 (inside: 1), inside vsync: 0
[-]   Hcount: 3718 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +235379ms, behind: 2083ms
[-]   Vblank: 13 (inside: 0), inside vsync: 0
[-]   Hcount: 3861 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +235379ms, behind: 16664ms
[-]   Vblank: 14 (inside: 1), inside vsync: 0
[-]   Hcount: 4004 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +252043ms, behind: 2083ms
[-]   Vblank: 14 (inside: 0), inside vsync: 0
[-]   Hcount: 4147 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +252043ms, behind: 16664ms
[-]   Vblank: 15 (inside: 1), inside vsync: 0
[-]   Hcount: 4290 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +268707ms, behind: 2083ms
[-]   Vblank: 15 (inside: 0), inside vsync: 0
[-]   Hcount: 4433 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +268707ms, behind: 16664ms
[-]   Vblank: 16 (inside: 1), inside vsync: 0
[-]   Hcount: 4576 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +285371ms, behind: 2083ms
[-]   Vblank: 16 (inside: 0), inside vsync: 0
[-]   Hcount: 4719 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +285371ms, behind: 16664ms
[-]   Vblank: 17 (inside: 1), inside vsync: 0
[-]   Hcount: 4862 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +302035ms, behind: 2083ms
[-]   Vblank: 17 (inside: 0), inside vsync: 0
[-]   Hcount: 5005 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +302035ms, behind: 16664ms
[-]   Vblank: 18 (inside: 1), inside vsync: 0
[-]   Hcount: 5148 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +318699ms, behind: 2083ms
[-]   Vblank: 18 (inside: 0), inside vsync: 0
[-]   Hcount: 5291 (+15)
[-] Peek inside interrupt: 00000040
[-]   Latest sample at: +318699ms, behind: 16664ms
[-]   Vblank: 19 (inside: 1), inside vsync: 0
[-]   Hcount: 5434 (+5)
[r] Read outside interrupt: 00000001
[-]   Latest sample at: +335363ms, behind: 2083ms
[-]   Vblank: 19 (inside: 0), inside vsync: 0
[-]   Hcount: 5577 (+15)
[r] Possible: 28
